### PR TITLE
feat(template): add dict, map and append functions

### PR DESCRIPTION
**Pull request by TheMeier**
These functions are useful for constructing complex data structures in templates. For example to compose  K/V arrays to use with the existing `toJSON` function. The dict function allows for creating maps from a list of key-value pairs, the list function creates slices from a list of values, and the append function allows for adding elements to a slice.

The names dict and list where chosen instead of map and slice because there is very widspread use of the sprig library in templates, (eg helm) and these names are used in that library.

#### Pull Request Checklist
Please check all the applicable boxes.

- Please list all open issue(s) discussed with maintainers related to this change
    - Fixes #<issue number>
    
- Is this a new Receiver integration?
    - [ ] I have already tried to use the [Webhook Receiver Integration](https://prometheus.io/docs/alerting/latest/configuration/#webhook_config) and [3rd party integrations](https://prometheus.io/docs/operating/integrations/#alertmanager-webhook-receiver) before adding this new Receiver Integration
- Is this a bugfix?
    - [ ] I have added tests that can reproduce the bug which pass with this bugfix applied
- Is this a new feature?
    - [x] I have added tests that test the new feature's functionality
- Does this change affect performance?
    - [ ] I have provided benchmarks comparison that shows performance is improved or is not degraded
        - You can use [`benchstat`](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to compare benchmarks
    - [ ] I have added new benchmarks if required or requested by maintainers
- Is this a breaking change?
    - [ ] My changes do not break the existing cluster messages
    - [ ] My changes do not break the existing api
- [x] I have added/updated the required documentation
- [x] I have signed-off my commits
- [x] I will follow [best practices for contributing to this project](https://docs.github.com/en/get-started/exploring-projects-on-github/contributing-to-open-source)

#### Which user-facing changes does this PR introduce?

```release-notes

```

## Summary by CodeRabbit

* **New Features**
  * Added three template helpers to build lists, append items, and construct key/value maps directly in templates.

* **Documentation**
  * Documented the new helpers with usage examples and expected error conditions.

* **Tests**
  * Added tests covering list/dict behaviors, nil handling, invalid-argument and non-string-key error cases, list mutation, and composed template scenarios.

### CODE REVIEW

File: template/template.go [hunk 1]
```diff
@@ -223,6 +224,28 @@ var DefaultFuncs = FuncMap{
 		}
 		return string(bytes), nil
 	},
+	"list": func(args ...any) ([]any, error) {
+		return args, nil
```
Comment (siavashs) [inline template/template.go:231]: We should return an empty slice if args is empty.
Example usage could be `list | toJson`.

```suggestion
		if args == nil {
			return []any{}, nil
		}
		return args, nil
```

File: docs/notifications.md [hunk 2]
```diff
@@ -85,14 +85,17 @@ templating.
...
 | match            | pattern, string            | [Regexp.MatchString](https://golang.org/pkg/regexp/#MatchString). Match a string using Regexp. |
 | reReplaceAll     | pattern, replacement, text | [Regexp.ReplaceAllString](http://golang.org/pkg/regexp/#Regexp.ReplaceAllString) Regexp substitution, unanchored. |
 | safeHtml         | text string                | [html/template.HTML](https://golang.org/pkg/html/template/#HTML), Marks string as HTML not requiring auto-escaping. |
 | safeUrl          | text string                | [html/template.URL](https://golang.org/pkg/html/template/#URL), Marks string as URL not requiring auto-escaping. |
 | since            | time.Time                  | [time.Since](https://pkg.go.dev/time#Since), returns the duration of how much time passed from the provided time till the current system time. |
+| list             | ...any                     | Returns the passed arguments as a slice of interfaces. |
```
Comment (siavashs) [inline docs/notifications.md]: This is out of alphabetical order.

**Review by siavashs — CHANGES_REQUESTED** [review]

File: docs/notifications.md (hunk 2 above)
Comment (TheMeier) [reply docs/notifications.md]: right, missed since I named it `slice` first ...

fixed

File: template/template.go (hunk 1 above)
Comment (TheMeier) [reply template/template.go:231]: changed as suggested

Comment (SoloJacobs) [reply template/template.go:231]: Silly question, but is this actually correct? I tried adding the following test without the nil check, and it still passes:

```go
   title: "Template using empty list piped to toJson",                                                                                                                                                                                     
   in:    `{{ list | toJson }}`,                                                                                                                                                                                                           
   exp:   `[]`,                                                                                                                                                                                                                            
```

It is true, that `json.Marshal` produces `null` if passed a `nil` pointer, but the template engine does not  function pass `nil`.

File: template/template.go [hunk 3]
```diff
@@ -223,6 +224,31 @@ var DefaultFuncs = FuncMap{
 		}
 		return string(bytes), nil
 	},
+	"list": func(args ...any) ([]any, error) {
```
Comment (SoloJacobs) [inline template/template.go:227]: Should we simplify the function signature here to ` func(args ...any) []any`?

File: template/template.go (hunk 1 above)
Comment (TheMeier) [reply template/template.go:231]: There is a tiny difference one comperes to nil the other doesen't
see https://go.dev/play/p/q2Ink2VYsNf

[sprig](https://github.com/Masterminds/sprig/blob/master/list.go#L14) doesn't care. But with the extra logic it is more correct.

File: template/template.go (hunk 3 above)
Comment (TheMeier) [reply template/template.go:227]: yes. updated

File: template/template_test.go [hunk 4]
```diff
@@ -313,7 +313,7 @@ func TestTemplateExpansion(t *testing.T) {
 		},
 		{
 			title: "Template with invalid syntax",
-			in:    `{{ `,
+			in:    `{{ foo }} `,
 			fail:  true,
```
[the remark this answers is not in the document]
Comment (TheMeier) [reply template/template_test.go:317]: removed change

File: template/template.go (hunk 1 above)
Comment (SoloJacobs) [reply template/template.go:231]: I think we are all aware how variadic functions behave, but that was not the point:

The template engine does not call `list`  (the go function) with no arguments, when list (the template function) is called with no arguments. This can be verified by removing the check and running the above test. So, the check doesn't change runtime behavior. We can keep it to make the intent explicit and guard against potential edge cases.
